Improve code coverage for System.Security.Cryptography.Xml

hootanht opened the pull request:
Fixes #20508

Add tests for `KeyInfoClause` and `CanonicalXmlEntityReference` to improve code coverage.

* **KeyInfoClauseTests.cs**
  - Add a new test class `KeyInfoClauseTests`.
  - Add a test method `LoadXml_ValidXml_Success` to test loading valid XML.
  - Add a test method `LoadXml_InvalidXml_ThrowsException` to test loading invalid XML.
  - Add a test method `GetXml_ReturnsExpectedXml` to test getting XML from `KeyInfoClause`.

* **CanonicalXmlEntityReferenceTests.cs**
  - Add a new test class `CanonicalXmlEntityReferenceTests`.
  - Add a test method `Write_WritesExpectedOutput` to test writing canonical XML.
  - Add a test method `WriteHash_WritesExpectedHash` to test writing hash of canonical XML.
  - Add a test method `IsInNodeSet_GetSet_ReturnsExpected` to test getting and setting `IsInNodeSet`.

hootanht committed:
Improve code coverage for System.Security.Cryptography.Xml

Fixes #20508

Add tests for `KeyInfoClause` and `CanonicalXmlEntityReference` to improve code coverage.

* **KeyInfoClauseTests.cs**
  - Add a new test class `KeyInfoClauseTests`.
  - Add a test method `LoadXml_ValidXml_Success` to test loading valid XML.
  - Add a test method `LoadXml_InvalidXml_ThrowsException` to test loading invalid XML.
  - Add a test method `GetXml_ReturnsExpectedXml` to test getting XML from `KeyInfoClause`.

* **CanonicalXmlEntityReferenceTests.cs**
  - Add a new test class `CanonicalXmlEntityReferenceTests`.
  - Add a test method `Write_WritesExpectedOutput` to test writing canonical XML.
  - Add a test method `WriteHash_WritesExpectedHash` to test writing hash of canonical XML.
  - Add a test method `IsInNodeSet_GetSet_ReturnsExpected` to test getting and setting `IsInNodeSet`.

vcsjones commented on src/libraries/System.Security.Cryptography.Xml/tests/CanonicalXmlEntityReferenceTests.cs:
```diff
@@ -0,0 +1,49 @@
+        {
+            var xmlDocument = new XmlDocument();
+            var entityReference = new CanonicalXmlEntityReference("entity", xmlDocument, true);
+            var hashAlgorithm = SHA256.Create();
```
This is not a place we would use `var`. Please see the [style guide](https://github.com/dotnet/runtime/blob/ffd3b187038a107e29cd425de3a748540a75e33e/docs/coding-guidelines/coding-style.md#L4), particularly point 10. So this should be explicitly typed.

Additionally, since `SHA256` is disposable, we generally try to make sure that we dispose of things, even in unit tests. Please put this in a `using`.

vcsjones commented on src/libraries/System.Security.Cryptography.Xml/tests/CanonicalXmlEntityReferenceTests.cs:
```diff
@@ -0,0 +1,49 @@
+
+            entityReference.WriteHash(hashAlgorithm, DocPosition.InsideRootElement, anc);
+
+            var expectedHash = hashAlgorithm.ComputeHash(Encoding.UTF8.GetBytes("<!ENTITY entity SYSTEM \"entity\">"));
```
Nit: another `var` that should not be here.

vcsjones commented on src/libraries/System.Security.Cryptography.Xml/tests/CanonicalXmlEntityReferenceTests.cs:
```diff
@@ -0,0 +1,49 @@
+
+namespace System.Security.Cryptography.Xml.Tests
+{
+    public class CanonicalXmlEntityReferenceTests
```
I don't see anything in this class that uses instance members besides methods. Can we mark this class as `static` (as well as all of the test methods)?

Same feedback to the other new file.

vcsjones commented on src/libraries/System.Security.Cryptography.Xml/tests/KeyInfoClauseTests.cs:
```diff
@@ -0,0 +1,67 @@
+        {
+            var xmlDocument = new XmlDocument();
+            xmlDocument.LoadXml("<KeyInfo><KeyName>Test</KeyName></KeyInfo>");
+            var keyInfoClause = new TestKeyInfoClause();
```
It's not very clear to me what we are testing here. All of these tests are testing a test class, so the tests are just testing `TestKeyInfoClause`.

The only thing on `KeyInfoClause` that actually has functionality to be tested is `GetXml(XmlDocument xmlDocument)` when it isn't overridden.

vcsjones reviewed (commented):
👋 thanks for the contribution!

The `dotnet/runtime` repository does not automatically include .cs files in projects, they must be explicitly added to the project file.

As a result, these tests are not actually running. Can you please add them to https://github.com/dotnet/runtime/blob/main/src/libraries/System.Security.Cryptography.Xml/tests/System.Security.Cryptography.Xml.Tests.csproj?

hootanht committed:
Add tests for `SignedXml.CheckSignature(X509Certificate2 certificate, bool verifySignatureOnly)`, `System.Security.Cryptography.Xml.KeyInfoClause`, and `System.Security.Cryptography.Xml.CanonicalXmlEntityReference` in `EncryptedXmlTest.cs`.

* **CheckSignature**: Add test for `CheckSignature` method with `X509Certificate2` and `verifySignatureOnly` parameter.
* **KeyInfoClause**: Add tests for `GetXml` and `LoadXml` methods.
* **CanonicalXmlEntityReference**: Add tests for `Write`, `WriteHash`, and `IsInNodeSet` methods.

Add `SymmetricKeyWrapTest.cs` to test `System.Security.Cryptography.Xml.SymmetricKeyWrap`.

Add `EncryptionMethodTest.cs` to test `System.Security.Cryptography.Xml.EncryptionMethod`.

hootanht committed:
Update EncryptedXmlTest.cs

hootanht committed:
Update EncryptedXmlTest.cs

hootanht committed:
Merge pull request #1 from hootanht/add-tests

Add tests for `SignedXml.CheckSignature` base on code review

hootanht committed:
Improve code coverage for System.Security.Cryptography.Xml

hootanht committed:
Add new test files to project file

Two new test files, `CanonicalXmlEntityReferenceTests.cs` and
`KeyInfoClauseTests.cs`, have been added to the project file
`System.Security.Cryptography.Xml.Tests.csproj`. These files are
included in the `<ItemGroup>` section, which lists the source files
to be compiled.

hootanht committed:
Refactor test methods to be static in CanonicalXmlEntityReferenceTests

Modified access modifiers of three test methods in `CanonicalXmlEntityReferenceTests.cs`:
- `Write_WritesExpectedOutput`
- `WriteHash_WritesExpectedHash`
- `IsInNodeSet_GetSet_ReturnsExpected`

Changed from instance methods (`public void`) to static methods (`public static void`).

hootanht committed:
Refactor to use 'var' keyword for local variables

Refactored `CanonicalXmlEntityReferenceTests.cs` and `KeyInfoClauseTests.cs` to replace explicit type declarations with `var` where the type is evident from the context. This change improves code readability and conciseness without altering functionality. Specifically, updated `XmlDocument`, `CanonicalXmlEntityReference`, `KeyInfoName`, and `XmlElement` variables.

hootanht committed:
Update CanonicalXmlEntityReferenceTests.cs

hootanht committed:
Update CanonicalXmlEntityReferenceTests.cs

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/tests/EncryptionMethodTest.cs:
```diff
@@ -0,0 +1,53 @@
+        public void GetXml_ReturnsExpectedXml()
+        {
+            var encryptionMethod = new EncryptionMethod("http://www.w3.org/2001/04/xmlenc#aes256-cbc");
+            var xmlElement = encryptionMethod.GetXml();
```
This use of `var` violates our style guidelines.

Never using `var` is always style-conforming in this repository.  If you think that a line like `EncryptionMethod encryptionMethod = new EncryptionMethod(...);` is too long, `var` is acceptable there (the type name is present on the line), but target-typed-new is probably the better choice over var (`EncryptionMethod encryptionMethod = new(...);`).

Which means that for new code, `var` would only really be used in a cast statement, if ever.

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/tests/System.Security.Cryptography.Xml.Tests.csproj:
```diff
@@ -55,6 +55,8 @@
     <Compile Include="XmlDsigXsltTransformTest.cs" />
     <Compile Include="XmlLicenseEncryptedRef.cs" />
     <Compile Include="XmlLicenseTransformTest.cs" />
+    <Compile Include="CanonicalXmlEntityReferenceTests.cs" />
+    <Compile Include="KeyInfoClauseTests.cs" />
   </ItemGroup>
```
Alphabetical, please :smile:

hootanht committed:
Update EncryptionMethodTest.cs

hootanht committed:
Update System.Security.Cryptography.Xml.Tests.csproj

vcsjones commented on src/libraries/System.Security.Cryptography.Xml/tests/CanonicalXmlEntityReferenceTests.cs:
```diff
@@ -0,0 +1,48 @@
+        public static void Write_WritesExpectedOutput()
+        {
+            var xmlDocument = new XmlDocument();
+            var entityReference = new CanonicalXmlEntityReference("entity", xmlDocument, true);
```
The tests are failing to compile now because `CanonicalXmlEntityReference` is `internal`. We don't typically unit test internal classes, only the public API surface.

vcsjones commented on src/libraries/System.Security.Cryptography.Xml/tests/CanonicalXmlEntityReferenceTests.cs:
```diff
@@ -0,0 +1,48 @@
+            var xmlDocument = new XmlDocument();
+            var entityReference = new CanonicalXmlEntityReference("entity", xmlDocument, true);
+            var stringBuilder = new StringBuilder();
+            var anc = new AncestralNamespaceContextManager();
```
Likewise, this type is `internal`.

vcsjones reviewed (commented):
As the tests are now showing, these tests do not compile because you are attempting to test and use `internal` types and members.

The unit test project here does not have access to internal members, and we very rarely make use of `InternalsVisibleTo`. `internal` things are typically tested by way of them being used from `public` things (behavioral tests).

I would also consider running tests locally before you submit them, or at least ensure they compile. The documentation on how to get started with local developer is in the [workflow instructions](https://github.com/dotnet/runtime/blob/main/docs/workflow/README.md).

At this point I would recommend identifying public types and functionality that are lacking test coverage.
